### PR TITLE
Enhancements to Artifactory Integration and NPM dependency tree building

### PR DESCRIPTION
**Pull request by eranturgeman**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Summary of Changes:

Automatically collects and inserts details of the Artifactory server and resolve repository into Audit parameters (for future use in dependency tree construction). This adjustment applies to all flows that construct a dependency tree.

Enhances the building of NPM dependency trees to resolve dependencies from an Artifactory server when configured by the user.

Introduces new parameters within Audit parameters to allow users to provide customized installation commands for building dependency trees. Note that not all entry points support the provision of these parameters.

Depends on: https://github.com/jfrog/build-info-go/pull/207

### CODE REVIEW

File: xray/commands/audit/sca/npm/npm.go [hunk 1]
```diff
@@ -31,6 +39,18 @@ func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils
 
 	treeDepsParam := createTreeDepsParam(params)
 
+	var restoreNpmrcFunc func() error
+	restoreNpmrcFunc, err = configureResolutionServerIfNeeded(params)
+	if err != nil {
+		log.Warn(fmt.Sprintf("Configuring an Artifactory server failed. Dependencies will be resolved from NPM default registry if needed\nFailure cause: %s", err.Error()))
+		err = nil
+	}
```
Comment (eyalbe4) [inline xray/commands/audit/sca/npm/npm.go:47]: I'm not sure that the code should decide whether to skip resolution from Artifactory based on an error.

File: xray/commands/audit/sca/npm/npm.go [hunk 2]
```diff
@@ -47,14 +67,78 @@ func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils
...
+				err = fmt.Errorf("couldn't get resolving server details: %s", err.Error())
+				return
+			}
+		}
+		if depsRepo == "" {
+			depsRepo = npmConfigYamlData.GetString("resolver.repo")
+		}
+	}
+
+	restoreNpmrcFunc, err = configNpmResolutionServer(depsRepo, serverDetails)
```
Comment (eyalbe4) [inline xray/commands/audit/sca/npm/npm.go:112]: Is this actually a restore function? What are we restoring here?

File: xray/commands/audit/sca/npm/npm.go [hunk 3]
```diff
@@ -47,14 +67,78 @@ func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils
...
+	if err != nil {
+		err = fmt.Errorf("configuring an artifactory server for resolution failed: %s", err.Error())
```
Comment (eyalbe4) [inline xray/commands/audit/sca/npm/npm.go:114]: Shouldn't we add add a return statement right after this line?

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: xray/commands/audit/sca/npm/npm.go [hunk 4]
```diff
@@ -47,14 +67,78 @@ func BuildDependencyTree(params utils.AuditParams) (dependencyTrees []*xrayUtils
...
+func configureResolutionServerIfNeeded(params utils.AuditParams) (restoreNpmrcFunc func() error, err error) {
+	depsRepo := params.DepsRepo()
+	serverDetails, err := params.ServerDetails()
+	if err != nil {
+		err = fmt.Errorf("couldn't get resolving server details: %s", err.Error())
+		return
+	}
+	if depsRepo == "" || serverDetails == nil {
+		// In case we don't have DepsRepo or ServerDetails, we search for a Jfrog config file, if exists, in order to get them
+		var isNpmConfigFileExists bool
```
[the remark this answers is not in the document]
Comment (eranturgeman) [reply xray/commands/audit/sca/npm/npm.go:81]: deprecated

File: xray/commands/audit/sca/npm/npm.go (hunk 2 above)
Comment (eranturgeman) [reply xray/commands/audit/sca/npm/npm.go:112]: We restore the state to the previous npmrc (or deleting it if wasn't exist) in order to not leaving any trace to out action after running 'restore' for audit.
This logic is taken from frogbot

File: xray/commands/audit/sca/npm/npm.go (hunk 3 above)
Comment (eranturgeman) [reply xray/commands/audit/sca/npm/npm.go:114]: fixed

File: xray/commands/audit/sca/npm/npm.go (hunk 1 above)
Comment (eranturgeman) [reply xray/commands/audit/sca/npm/npm.go:47]: deprecated

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Looks good.
Let's just improve the grammar of the inline comments, PR title and PR description before merging.

**Pull request merged**